File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head) {
            return nullptr; 
        }
        ListNode* lastnode = new ListNode(head->val); 
        ListNode* curr = lastnode; 
        head = head -> next; 
        while ( head) {
            ListNode* firstnode = new ListNode(head->val, curr); 
            curr = firstnode; 
            ListNode* delnode = head; 
            head = head -> next; 
            delete delnode; 
            
        }
        return curr;  
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]'):

        if not head:
            return None

        curr = head 

        NewHead = Node(head.val)
        NewHead_curr = NewHead

        d = {}

        d[curr] = NewHead_curr

        while curr.next: 
            curr = curr.next 
            NewHead_curr.next = Node(curr.val)

            NewHead_curr = NewHead_curr.next
            d[curr] = NewHead_curr

        NewHead_curr = NewHead
        curr = head 
        
        while NewHead_curr: 
            if curr.random:
                NewHead_curr.random = d[curr.random]
            
            NewHead_curr = NewHead_curr.next 
            curr = curr.next 
        
        return NewHead
        
    


        
        